File: php/06_facade/main.php
<?php

class CPU {
    public function start() {
        echo "\nCPU - staring\n";
    }

    public function execute() {
        echo "\nCPU - executing\n";
    }
}

class Memory {
    public function load() {
        echo "\nMemory - loading\n";
    }
}

class Disk {
    public function read() {
        echo "\nDisk - reading\n";
    }
}

class ComputerFacade {
    private $cpuClass, $memoryClass, $diskClass;

    public function __construct() {
        $this->cpuClass = new CPU();
        $this->memoryClass = new Memory();
        $this->diskClass = new Disk();
    }

    public function start() {
        $this->cpuClass->start();
        $this->memoryClass->load();
        $this->diskClass->read();
        $this->cpuClass->execute();
    }
}

$facade = new ComputerFacade();

$facade->start();

?>

File: go/06_facade/main.go
package main

import "fmt"

type CPU struct{}

func (c *CPU) start() {
	fmt.Println("CPU starting")
}

func (c *CPU) execute() {
	fmt.Println("CPU executing")
}

type Memory struct{}

func (m *Memory) load() {
	fmt.Println("Memory loaded")
}

type Disk struct{}

func (d *Disk) read() {
	fmt.Println("Disk reading")
}

// facade
type ComputerFacade struct {
	cpu    CPU
	memory Memory
	disk   Disk
}

func NewComputerFacade() *ComputerFacade {
	return &ComputerFacade{
		cpu:    CPU{},
		memory: Memory{},
		disk:   Disk{},
	}
}

func (cf *ComputerFacade) start() {
	cf.cpu.start()
	cf.memory.load()
	cf.disk.read()
	cf.cpu.execute()
}

func main() {
	cf := NewComputerFacade()

	cf.start()
}
